queue.go

```diff
@@ -1,56 +1,222 @@
-// interface `Job`
 package bqueue
 
-import "sync"
+import (
+	"context"
+	"fmt"
+	"log"
+	"runtime"
+	"sync"
+)
 
-var WG sync.WaitGroup
+const (
+	// DefaultLimit is the default job queue limit.
+	DefaultLimit = 128
 
-// Queue that process jobs reveived
+	// UnlimitedWorkers can passed to Workers when run in dynamic
+	// mode to use an unlimited amount of workers.
+	UnlimitedWorkers = -1
+)
+
+// Queue processes jobs.
 type Queue struct {
-	maxWorker   int
-	JobRequests chan chan Job
-	JobReceived chan Job
+	limiter chan struct{}
+	workers int
+	limit   int
+	run     func() error
+	jobs    chan Job
+	log     Logger
+	wg      sync.WaitGroup
 }
 
-// New Queue object
-// @param MaxWorkerint is the maximum job at a single time that can be handled
-func New(maxWorker int) *Queue {
-	JobRequests := make(chan chan Job, maxWorker)
-	return &Queue{
-		maxWorker:   maxWorker,
-		JobRequests: JobRequests,
-		JobReceived: make(chan Job, 128),
+// Option represents a queue option.
+type Option func(q *Queue) error
+
+// Workers sets the number of workers which will process jobs from the queue.
+// The constant UnlimitedWorkers can be used along side dynamic workers, the default,
+// which will enable unlimited workers.
+// Default is runtime.NumCPU() workers.
+func Workers(count int) Option {
+	return func(q *Queue) error {
+		if count < UnlimitedWorkers {
+			return fmt.Errorf("workers must be at least %d, got %d", UnlimitedWorkers, count)
+		}
+
+		q.workers = count
+
+		return nil
 	}
 }
 
-// Start the queue
-func (q *Queue) Start() {
-	for i := 0; i < q.maxWorker; i++ {
-		id := i + 1
-		worker := newWorker(id, q.JobRequests)
-		worker.start()
+// Limit sets the maximum number of jobs queued before an error is returned.
+// Default is DefaultLimit maximum queued jobs..
+func Limit(count int) Option {
+	return func(q *Queue) error {
+		if count < 0 {
+			return fmt.Errorf("limit must be at least 0, got %d", count)
+		}
+
+		q.limit = count
+
+		return nil
 	}
+}
 
-	go q.dispatch()
+// Log sets the logger for Queue.
+// Default is log.Default().
+func Log(l Logger) Option {
+	return func(q *Queue) error {
+		q.log = l
+		return nil
+	}
 }
 
-// CollectJob Adds a job to the Queue
-func (q *Queue) CollectJob(job Job) {
-	WG.Add(1)
-	q.JobReceived <- job
+// Static configures the Queue to use a static number of pre-spawned
+// workers instead of dynamic workers, which can be beneficial for
+// inexpensive jobs.
+// Default behaviour is to use dynamic workers.
+func Static() Option {
+	return func(q *Queue) error {
+		q.run = q.static
+		return nil
+	}
 }
 
-func (q *Queue) dispatch() {
-	for {
-		select {
-		case job := <-q.JobReceived:
-			go func() {
-				jobRequest := <-q.JobRequests
-				jobRequest <- job
-			}()
+// New creates a fully initialised Queue with the given options.
+func New(options ...Option) (*Queue, error) {
+	q := &Queue{
+		workers: runtime.NumCPU(),
+		limit:   DefaultLimit,
+		log:     log.Default(),
```

Comment:
we should add `go 1.16` to the go.mod 

since that function was not allowed before